zetta_utils/augmentations/misalign.py

```diff
@@ -0,0 +1,165 @@
+from __future__ import annotations
+
+import math
+import random
+from typing import Any, Literal, TypeVar
+
+import attrs
+import torch
+from typeguard import typechecked
+
+from zetta_utils import builder
+from zetta_utils.geometry.vec import Vec3D
+from zetta_utils.layer.tools_base import JointIndexDataProcessor
+from zetta_utils.layer.volumetric.index import VolumetricIndex
+
+T = TypeVar("T", torch.Tensor, dict[str, Any])
+
+
+@builder.register("SlipMisalign")
+@typechecked
+@attrs.mutable
+class MisalignProcessor(JointIndexDataProcessor[T, VolumetricIndex]):
+    """
+    Minimum and maximum displacement is specified in unit.
+    The selected displacement will be rounded to always be divisible
+    by `disp_in_unit_must_be_divisible_by`, which should correspond to
+    the lowest resolution of the layers being cropped.
+
+    Data handling is implemented through recording the fraction
+    of the misalignment relative to the total bbox size.
+    """
+
+    prob: float
+    disp_min_in_unit: float
+    disp_max_in_unit: float
+    disp_in_unit_must_be_divisible_by: float | None = None
+    mode: Literal["slip", "step"] = "slip"
+    keys_to_apply: list[str] | None = None
+
+    prepared_disp_fraction: tuple[float, float] | None = attrs.field(init=False, default=None)
+
+    def process_index(
+        self, idx: VolumetricIndex, mode: Literal["read", "write"]
+    ) -> VolumetricIndex:
+        if mode != "read":
+            raise NotImplementedError()
+        disp_x_in_unit_magn = random.uniform(self.disp_min_in_unit, self.disp_max_in_unit)
+        disp_y_in_unit_magn = random.uniform(self.disp_min_in_unit, self.disp_max_in_unit)
+        if self.disp_in_unit_must_be_divisible_by is not None:
+            disp_x_in_unit_magn = math.floor(
```

Comment:
I'm totally confused here. If the purpose of `disp_in_unit_must_be_divisible_by` is as the name of the variable implies, then shouldn't it be
```
disp_x_in_unit_magn = (
        disp_x_in_unit_magn // self.disp_in_unit_must_be_divisible_by
    ) * self.disp_in_unit_must_be_divisible_by
```
?